akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/SlotState.scala

```diff
@@ -0,0 +1,195 @@
+      ???
+    }
+  }
+
+  // FIXME: unify with Connecting
+  case object PreConnecting extends ConnectedState with IdleState with WithRequestDispatching {
+    override def onConnectedAttemptSucceeded(ctx: SlotContext, outgoingConnection: Http.OutgoingConnection): SlotState = {
+      ctx.debug("Slot connection was (pre-)established")
+      Idle
+    }
+    override def onConnectionAttemptFailed(ctx: SlotContext, cause: Throwable): SlotState = {
+      ctx.debug("Connection attempt failed.")
+      // FIXME: register failed connection attempt, schedule request for rerun, backoff new connection attempts
+      ???
+    }
+
+    // FIXME (the problem here is that we install the completion handlers here and then another time in Connecting
+    //override def onDispatchRequest(ctx: SlotContext, requestContext: RequestContext): SlotState =
+    //???
+  }
+  final case class WaitingForEndOfRequestEntity(ongoingRequest: RequestContext) extends ConnectedState with BusyState {
+    override def onRequestEntityCompleted(ctx: SlotContext): SlotState = WaitingForResponse(ongoingRequest)
+
+    override def onConnectionFailed(ctx: SlotContext, cause: Throwable): SlotState = {
+      ctx.dispatchFailure(ongoingRequest, cause)
+      ctx.closeConnection()
+
+      Unconnected
+    }
+
+  }
+  final case class WaitingForResponse(ongoingRequest: RequestContext) extends ConnectedState with BusyState {
+    override def onResponseReceived(ctx: SlotContext, response: HttpResponse): SlotState = {
+      ctx.dispatchResponse(ongoingRequest, response)
+
+      WaitingForEndOfResponseEntity(ongoingRequest, response)
+    }
+
+    override def onConnectionFailed(ctx: SlotContext, cause: Throwable): SlotState = {
+      ctx.dispatchFailure(ongoingRequest, cause)
+      ctx.closeConnection()
+
+      Unconnected
+    }
+  }
+  /*case class WaitingForResponseEntitySubscription extends ConnectedState with BusyState {
+
+}*/
+  final case class WaitingForEndOfResponseEntity(
+    ongoingRequest:  RequestContext,
+    ongoingResponse: HttpResponse) extends ConnectedState with BusyState {
+
+    override def onResponseEntityCompleted(ctx: SlotContext): SlotState =
+      if (ctx.willCloseAfter(ongoingResponse) || ctx.isConnectionClosed) {
+        ctx.closeConnection()
+        Unconnected
+      } else
+        Idle
+
+    override def onResponseEntityFailed(ctx: SlotContext, cause: Throwable): SlotState = {
+      ctx.debug("Response entity failed with {}", cause)
+      // we cannot fail the response at this point, the response has already been dispatched
+      ctx.closeConnection()
+      Unconnected
+    }
+
+    // we ignore these signals here and expect that it will also be flagged on the entity stream
+    // FIXME: should we still add timeouts for these cases?
+    override def onConnectionFailed(ctx: SlotContext, cause: Throwable): SlotState = this
+    override def onConnectionCompleted(ctx: SlotContext): SlotState = this
+  }
+
+}
```

Comment:
great way to model these 👍  It's super clear what causes what now

Reply:
I agree, god design this!